### PR TITLE
Made frequency determination code stateful, implemented a virtual ring buffer

### PR DESCRIPTION
**Pull request by ewtoombs**
Yeah, just see logs.

### CODE REVIEW

File: arm/capture.c [hunk 1]
```diff
@@ -7,10 +7,12 @@
 #include <asoundlib.h>
 
 #include "capture.h"
-#include "freq.h"
 
 
 #define EXCRUCIATING_DETAIL 0
+#define LATENCY 10e-3f
+// It would just be stupid to run with a period any lower than this.
```
Comment (reinderien) [inline arm/capture.c]: Can you elaborate on this, rather than just saying stupid?

File: arm/capture.c [hunk 2]
```diff
@@ -760,13 +751,14 @@ static snd_pcm_sframes_t capture_wait(CaptureContext *restrict ctx)
 }
 
 
-void capture_period(
+void capture_capture_period(
```
Comment (reinderien) [inline arm/capture.c]: This method name is confusing. Can we choose something else? Perhaps `do_capture` here, and `get_capture_period` below?

File: arm/data/track [hunk 3]
```diff
@@ -0,0 +1,4 @@
+This is here so that this directory, arm/data, gets added to the git repo,
+since git will not track empty directories. This is where any data generated
+by pianotuner as it runs gets saved. Right now, it's just debug data, like
+autocorrelation snapshots.
```
Comment (reinderien) [inline arm/data/track:4]: OK. Why worry about having an empty dir in the repo at all? If the dir is missing, the application should create it on start.

File: arm/freq.c [hunk 4]
```diff
@@ -7,268 +7,97 @@
...
+//f: 012.4...8...
+//   .   .   0===
+//   .   .  1===-
+//   .   . 2-==--
+//   .   .3--=---
+//   .   4-------
+//   .  5--- ----
+//   . 6---  ----
+//   .7---   ----
+//           ^f0
```
Comment (reinderien) [inline arm/freq.c]: Could we `/* */` this for my sanity?

File: arm/util.h [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#include <stdlib.h>
+
+// determines the length  of a static array.
```
Comment (reinderien) [inline arm/util.h]: We can lose a double-space here

File: arm/vrb.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        0  // offset within file
+    );
+    assert(b->mem != MAP_FAILED);
+
+    // mmap can not arbitrarily remap memory regions to other regions by
+    // address. It can only produce shared mappings through the fd of a shared
+    // file. So, we need to make a fake file backed by memory only and use its
+    // fd. This is memfd_create's purpose.
+    // CLOEXEC is not on by default, but is a good default, lol. Just look it
+    // up in man 2 memfd_create for more.
```
Comment (reinderien) [inline arm/vrb.c]: Can we `/**/` these blocks as well?

File: arm/vrb_test.c [hunk 7]
```diff
@@ -0,0 +1,76 @@
+#include <inttypes.h>
```
Comment (reinderien) [inline arm/vrb_test.c:1]: YAY TESTS

File: arm/vrb_test.c [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+typedef uint32_t u32;
+typedef uint64_t u64;
+typedef int32_t i32;
+typedef int64_t i64;
```
Comment (reinderien) [inline arm/vrb_test.c]: I don't think these are worth defining.

**Review by reinderien — CHANGES_REQUESTED** [review]
Very exciting!

File: arm/capture.c (hunk 1 above)
Comment (ewtoombs) [reply arm/capture.c]: Or I could just nuke it. A somewhat high latency automatically forbids ridiculous capture periods.

File: arm/data/track (hunk 3 above)
Comment (ewtoombs) [reply arm/data/track:4]: It was easier to make this file than to figure out how test for missing directories then make directories using the C standard library. Like a /lot/ easier. And this is only for debugging anyway.

File: arm/vrb_test.c (hunk 7 above)
Comment (ewtoombs) [reply arm/vrb_test.c:1]: Yeah, I was really doubting my mmap magic would work at all. I had to see it with my own eyes. And with a test.

File: arm/vrb_test.c (hunk 8 above)
Comment (ewtoombs) [reply arm/vrb_test.c]: Those may or may not have been copy pasted in from one of my C test templates lol

**Pull request merged**